File: server/src/main/com/broll/gainea/server/core/cards/impl/direct/C_Pirates.java
package com.broll.gainea.server.core.cards.impl.direct;

import com.broll.gainea.misc.RandomUtils;
import com.broll.gainea.server.core.cards.DirectlyPlayedCard;
import com.broll.gainea.server.core.map.Ship;
import com.broll.gainea.server.core.objects.Soldier;
import com.broll.gainea.server.core.utils.LocationUtils;
import com.broll.gainea.server.core.utils.UnitControl;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class C_Pirates extends DirectlyPlayedCard {
    public C_Pirates() {
        super(38, "Piraten!", "Piraten befallen zufällige unbesetzte Schiffe");
        setDrawChance(0.8f);
    }

    @Override
    protected void play() {
        List<Ship> freeShips = game.getMap().getAllShips().stream().filter(LocationUtils::noControlledUnits).collect(Collectors.toList());
        Collections.shuffle(freeShips);
        int count = RandomUtils.random(2, 3);
        for (int i = 0; i < Math.min(count, freeShips.size()); i++) {
            Soldier pirate = new Soldier(null);
            pirate.setName("Pirat");
            pirate.setIcon(121);
            pirate.setPower(2);
            pirate.setHealth(2);
            UnitControl.spawn(game, pirate, freeShips.get(i));
        }
    }

}

File: server/src/main/com/broll/gainea/server/core/cards/impl/direct/C_Pirates.kt
package com.broll.gainea.server.core.cards.impl.direct

import com.broll.gainea.misc.RandomUtils
import com.broll.gainea.server.core.cards.DirectlyPlayedCard
import com.broll.gainea.server.core.objects.Soldier
import com.broll.gainea.server.core.utils.UnitControl.spawn
import com.broll.gainea.server.core.utils.noPlayerUnits

class C_Pirates : DirectlyPlayedCard(38, "Piraten!", "Piraten befallen zufällige unbesetzte Schiffe") {
    init {
        drawChance = 0.8f
    }

    override fun play() {
        val freeShips = game.map.allShips.filter { it.noPlayerUnits() }.shuffled()
        val count = RandomUtils.random(2, 4)
        for (i in 0 until Math.min(count, freeShips.size)) {
            val pirate = Soldier(game.neutralPlayer)
            pirate.name = "Pirat"
            pirate.icon = 121
            pirate.setPower(2)
            pirate.setHealth(2)
            game.spawn(pirate, freeShips[i])
        }
    }
}
